### DISCUSSION TITLE
Use MSW in same process as cypress

### DISCUSSION THREAD

**Discussion by knoefel** [cypress-io/cypress#22265] [General]
First of all thank you for this great tool for creating e2e tests!!!

My setup is the following:
- remix with [msw.js](https://github.com/mswjs/msw)
- Managing mocks with a [mswjs/data](https://github.com/mswjs/data) factory
- cypress 10

What i want is to be able to initialise, update or reset the mock factory from within the cypress tests. So for example to test this use case: 

User navigates to a list, then adds an item to that list, afterwards the list should contain that item. After the test i want to reset the mock factory to have a clean state for the next test. 

The problem is that `msw` and `mswjs/data` run in a different process than `cypress`, so i can't use the factory methods. I tried to run `msw` within `cypress` but get an error that it can't be executed within a browser environment. As i need to mock `remix` server code with `msw` i need to use `msw`'s `setupServer` instead of `setupWorker` (for the browser api mocking).

Any ideas how i can get this to work? Any tip would be awesome as i'm really stuck right now.

Thanks in advance!!!

**Comment by singhAmandeep007** [comment]
@knoefel 

I have prepared a React + Vite boilerplate which includes integration of MSW with Vitest and Cypress.

This allows to easily test an application with mocked responses.
With mocked response you can easily simulate a success scenario as well as failure scenario, thus covering most of user stories.

What's included ?
✅ Automatic reset
✅ Isolation
✅ Logging
✅ Fully Type safe
✅ With middleware

[Repo](https://github.com/singhAmandeep007/react-vite-boilerplate)
[MSW + Cypress](https://github.com/singhAmandeep007/react-vite-boilerplate/blob/main/cypress/support/msw.ts)
[MSW + Vitest](https://github.com/singhAmandeep007/react-vite-boilerplate/blob/main/src/tests/testServer.ts)

Please give it a ⭐️ if it helped.

**Reply by nickyhaze** [reply]
> @knoefel
> 
> I have prepared a React + Vite boilerplate which includes integration of MSW with Vitest and Cypress.
> 
> This allows to easily test an application with mocked responses. With mocked response you can easily simulate a success scenario as well as failure scenario, thus covering most of user stories.
> 
> What's included ? ✅ Automatic reset ✅ Isolation ✅ Logging ✅ Fully Type safe ✅ With middleware
> 
> [Repo](https://github.com/singhAmandeep007/react-vite-boilerplate) [MSW + Cypress](https://github.com/singhAmandeep007/react-vite-boilerplate/blob/main/cypress/support/msw.ts) [MSW + Vitest](https://github.com/singhAmandeep007/react-vite-boilerplate/blob/main/src/tests/testServer.ts)
> 
> Please give it a ⭐️ if it helped.

Do you have an example for using cypress headless and msw using setupServer?

**Comment by nickyhaze** [comment]
What about this approach?

https://github.com/mswjs/examples/pull/134

**Reply by MikeMcC399** [reply]
@nickyhaze 

> https://github.com/mswjs/examples/pull/134

I can't comment on the details, however the versions used in this PR are outdated. Node.js 20 will enter end-of-life in a few weeks.
Cypress 12 was last released in August 2023. The current version is Cypress 15.x. pnpm 9 is no longer current. pnpm 10 is the current version. Just to point out the versions I am more familiar with.

**Reply by nickyhaze** [reply]
Ah yes, I followed the versions that the other examples use

**Reply by MikeMcC399** [reply]
It looks like the other examples were updated 3 years ago
